src/verify_cert.rs

```diff
@@ -361,25 +361,20 @@ impl ExtendedKeyUsage {
                 }
                 Ok(())
             }
-            None => {
-                if matches!(self, Self::Required(_)) {
-                    return Err(Error::RequiredEkuNotFound);
-                }
-                // http://tools.ietf.org/html/rfc6960#section-4.2.2.2:
-                // "OCSP signing delegation SHALL be designated by the inclusion of
-                // id-kp-OCSPSigning in an extended key usage certificate extension
-                // included in the OCSP response signer's certificate."
-                //
-                // A missing EKU extension generally means "any EKU", but it is
-                // important that id-kp-OCSPSigning is explicit so that a normal
-                // end-entity certificate isn't able to sign trusted OCSP responses
-                // for itself or for other certificates issued by its issuing CA.
-                if self.key_purpose_id_equals(EKU_OCSP_SIGNING.oid_value) {
-                    return Err(Error::RequiredEkuNotFound);
-                }
-
-                Ok(())
-            }
+            // http://tools.ietf.org/html/rfc6960#section-4.2.2.2:
+            // "OCSP signing delegation SHALL be designated by the inclusion of
+            // id-kp-OCSPSigning in an extended key usage certificate extension
+            // included in the OCSP response signer's certificate."
+            //
+            // A missing EKU extension generally means "any EKU", but it is
+            // important that id-kp-OCSPSigning is explicit so that a normal
+            // end-entity certificate isn't able to sign trusted OCSP responses
+            // for itself or for other certificates issued by its issuing CA.
+            (None, Self::RequiredIfPresent(expected)) => match expected == &EKU_OCSP_SIGNING {
+                true => Err(Error::RequiredEkuNotFound),
+                false => Ok(()),
+            },
```

Comment:
if i'm reading this correctly, we could discard this special case and replace it with:

```
impl ExtendedKeyUsage {
    /// (the big comment)
    pub(crate) fn ocsp_issuance() -> Self { Self::Required(EKU_OCSP_SIGNING) }
}
```

(but also worth noting that there are no tests or callers for this behaviour.)

Reply:
I reworked the API to provide a more misuse-resistant API, please take a look.